File: java-gdb-activities-20/activity17/src/com/gdb/command/TransferCommand.java
package com.gdb.command;

import com.gdb.domain.IAccount;
import com.gdb.domain.Transaction;
import com.gdb.service.TransferService;

public class TransferCommand implements TransactionCommand {
    private static final long serialVersionUID = 1L;

    // ============================================================
    // 📝 STEP 3: Declare Fields
    //
    // INSTRUCTIONS:
    //   1. private IAccount fromAccount;
    //   2. private IAccount toAccount;
    //   3. private double amount;
    //   4. private int pin;
    //   5. private Transaction transaction;
    // ============================================================
    // TODO: declare fromAccount, toAccount, amount, pin, and transaction fields

    // ============================================================
    // 📝 STEP 4: Write Constructor
    //
    // INSTRUCTIONS:
    //   Accept (IAccount from, IAccount to, double amount, int pin); store all.
    // ============================================================
    // TODO: implement constructor
    public TransferCommand(IAccount from, IAccount to, double amount, int pin) {
        // TODO: Step 4 - implement constructor
    }

    // ============================================================
    // 📝 STEP 5: Implement execute()
    //
    // INSTRUCTIONS:
    //   1. Call new TransferService().transferWithTransaction(fromAccount, toAccount, amount, pin).
    //   2. Store returned Transaction in this.transaction.
    // ============================================================
    // TODO: implement execute() method
    @Override
    public void execute() throws Exception {
        // TODO: Step 5 - implement execute()
    }

    // ============================================================
    // 📝 STEP 6: Implement getTransaction()
    //
    // INSTRUCTIONS:
    //   Return this.transaction.
    // ============================================================
    // TODO: return this.transaction
    @Override
    public Transaction getTransaction() {
        // TODO: Step 6 - return this.transaction
        return null;
    }
}

File: python-gdb-activities-20/activity19-solution/gdb/command/transfer_command.py
# gdb/command/transfer_command.py
from typing import Optional
from datetime import datetime
from gdb.domain.iaccount import IAccount
from gdb.domain.transaction import Transaction
from gdb.domain.transaction_type import TransactionType
from gdb.command.transaction_command import TransactionCommand
from gdb.service.transfer_service import TransferService

class TransferCommand(TransactionCommand):
    """Command executing a funds transfer operation."""

    def __init__(self, transfer_service: TransferService, from_account: IAccount, to_account: IAccount, amount: float, pin: int) -> None:
        self._transfer_service = transfer_service
        self._from_account = from_account
        self._to_account = to_account
        self._amount = float(amount)
        self._pin = int(pin)
        self._transaction: Optional[Transaction] = None

    def execute(self) -> None:
        self._transfer_service.transfer(self._from_account, self._to_account, self._amount, self._pin)
        self._transaction = Transaction(
            transaction_id=Transaction.generate_id(),
            timestamp=datetime.now(),
            account_number=self._from_account.get_account_number(),
            transaction_type=TransactionType.TRANSFER_OUT,
            amount=self._amount,
            balance_after=self._from_account.get_balance(),
            status="SUCCESS",
            description=f"Transfer to #{self._to_account.get_account_number()}",
            from_account=self._from_account.get_account_number(),
            to_account=self._to_account.get_account_number(),
        )

    def get_transaction(self) -> Optional[Transaction]:
        return self._transaction

    def getTransaction(self) -> Optional[Transaction]:
        return self.get_transaction()
